File: Code/pit_hd110067_sim/hd110067_adaptive.py
# hd110067_adaptive.py
# PIT HD 110067 Simulation Engine (Adaptive / Process Fractal) — Python port
#
# Mirrors hd110067_adaptive.jl + Addendum v0.1 refinements
# - Φ planets in AU / days
# - K resonant modes as complex phasors
# - Adaptive alpha(t), mu(t)
# - dt-consistent gravity + PIT torque
#
# Outputs:
#   hd110067_adaptive_trace_py.csv
#   hd110067_adaptive_trace_py.png

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

G = 2.95912208286e-4  # AU^3 / (Solar Mass * Day^2)


# -----------------------------
# Structures
# -----------------------------

class Planet:
    def __init__(self, pid, mass, pos, vel, name):
        self.id = pid
        self.mass = float(mass)
        self.pos = np.array(pos, dtype=float)  # shape (3,)
        self.vel = np.array(vel, dtype=float)  # shape (3,)
        self.name = name


class KernelField:
    def __init__(
        self,
        modes,
        mu=0.9,
        nu=0.05,
        alpha=1e-5,
        k_alpha=1e-4,
        k_mu=1e-5,
        d_target=0.10,
        alpha_bounds=(1e-6, 0.2),
        mu_bounds=(0.01, 1.0),
    ):
        self.modes = np.array(modes, dtype=np.complex128)
        self.mu = float(mu)
        self.nu = float(nu)
        self.alpha = float(alpha)
        self.k_alpha = float(k_alpha)
        self.k_mu = float(k_mu)
        self.d_target = float(d_target)
        self.alpha_bounds = alpha_bounds
        self.mu_bounds = mu_bounds


class ResonanceRule:
    def __init__(self, inner_id, outer_id, p, q):
        self.inner_id = inner_id  # 1-based
        self.outer_id = outer_id  # 1-based
        self.p = int(p)
        self.q = int(q)


# -----------------------------
# Helpers
# -----------------------------

def get_lambda(p: Planet):
    """Mean longitude approximation."""
    return np.arctan2(p.pos[1], p.pos[0])


def measure_phasors(planets, rules):
    """Interface measurement: z_i = exp(i * phi_i)."""
    z = np.empty(len(rules), dtype=np.complex128)
    for i, rule in enumerate(rules):
        li = get_lambda(planets[rule.inner_id - 1])
        lo = get_lambda(planets[rule.outer_id - 1])
        phi = rule.p * lo - rule.q * li
        z[i] = np.exp(1j * phi)
    return z


# -----------------------------
# K-field update (memory evolution)
# -----------------------------

def update_kernel(K: KernelField, z, dt):
    """
    Discrete Euler-Lagrange for K:
      K <- K + eta*(z-K)*dt
    where eta = nu*(1-saturation).
    Returns dissonance D and saturation.
    """
    saturation = np.abs(np.mean(K.modes))
    eta = K.nu * (1.0 - saturation)

    diff = z - K.modes
    D = np.mean(np.abs(diff))

    K.modes = K.modes + eta * diff * dt

    # renormalize each mode to keep |K|<=1
    mags = np.abs(K.modes)
    too_big = mags > 1.0
    if np.any(too_big):
        K.modes[too_big] /= mags[too_big]

    return D, saturation


# -----------------------------
# Adaptive drift (Addendum v0.1)
# -----------------------------

def adapt_parameters(K: KernelField, D, dt):
    """
    Adaptive alpha + mu:
      alpha_dot = k_alpha*(D - D*)
      mu_dot    = -k_mu*D
    """
    K.alpha += K.k_alpha * (D - K.d_target) * dt
    K.mu    -= K.k_mu * D * dt

    K.alpha = float(np.clip(K.alpha, *K.alpha_bounds))
    K.mu    = float(np.clip(K.mu,    *K.mu_bounds))


# -----------------------------
# Φ update: gravity + PIT torque
# -----------------------------

def apply_forces(planets, star_mass, K: KernelField, rules, dt):
    # Gravity
    for p in planets:
        r = p.pos
        d = np.linalg.norm(r)
        acc = -G * star_mass * r / (d**3)
        p.vel += acc * dt

    # PIT torque
    if K.alpha > 1e-9:
        z = measure_phasors(planets, rules)
        zhat = np.array([0.0, 0.0, 1.0])  # axis for tangential direction

        for i, rule in enumerate(rules):
            k = K.modes[i]
            zi = z[i]

            phase_error = np.angle(k) - np.angle(zi)
            torque = K.alpha * np.abs(k) * np.sin(phase_error)

            pin = planets[rule.inner_id - 1]
            pout = planets[rule.outer_id - 1]

            tin3 = np.cross(zhat, pin.pos)
            tout3 = np.cross(zhat, pout.pos)

            tin = tin3 / (np.linalg.norm(tin3) + 1e-12)
            tout = tout3 / (np.linalg.norm(tout3) + 1e-12)

            pin.vel  += (torque / pin.mass)  * tin  * dt
            pout.vel -= (torque / pout.mass) * tout * dt


def step_positions(planets, dt):
    for p in planets:
        p.pos += p.vel * dt


# -----------------------------
# Main simulation
# -----------------------------

def run_sim(seed=123, dt=0.1, steps=20000, log_every=20):
    star_mass = 0.81
    m_scale = 3.003e-6

    radii = [0.0793, 0.1039, 0.1364, 0.1790, 0.2166, 0.2621]
    planets = [
        Planet(i+1, 5.69*m_scale, [radii[i], 0, 0], [0,0,0], chr(ord("b")+i))
        for i in range(6)
    ]

    rng = np.random.default_rng(seed)
    for p in planets:
        r = np.linalg.norm(p.pos)
        theta = rng.random() * 2*np.pi
        v = np.sqrt(G * star_mass / r)
        p.pos = np.array([r*np.cos(theta), r*np.sin(theta), 0.0])
        p.vel = np.array([-v*np.sin(theta), v*np.cos(theta), 0.0])

    rules = [
        ResonanceRule(1,2,3,2),
        ResonanceRule(2,3,3,2),
        ResonanceRule(3,4,3,2),
        ResonanceRule(4,5,4,3),
        ResonanceRule(5,6,4,3),
    ]

    init_modes = 0.01 * np.exp(1j*2*np.pi*rng.random(len(rules)))
    K = KernelField(
        init_modes,
        mu=0.90, nu=0.05, alpha=1e-5,
        k_alpha=1e-4, k_mu=1e-5,
        d_target=0.10
    )

    trace = {
        "Step": [], "Phi_Angle": [], "K_Strength": [],
        "Alpha": [], "Mu": [], "Dissonance": [], "Saturation": []
    }

    for t in range(1, steps+1):
        z = measure_phasors(planets, rules)
        D, sat = update_kernel(K, z, dt)
        adapt_parameters(K, D, dt)

        apply_forces(planets, star_mass, K, rules, dt)
        step_positions(planets, dt)

        if t % log_every == 0:
            trace["Step"].append(t)
            trace["Phi_Angle"].append(np.angle(z[0]))
            trace["K_Strength"].append(np.abs(K.modes[0]))
            trace["Alpha"].append(K.alpha)
            trace["Mu"].append(K.mu)
            trace["Dissonance"].append(D)
            trace["Saturation"].append(sat)

    return pd.DataFrame(trace)


def main():
    df = run_sim()

    csv_path = "hd110067_adaptive_trace_py.csv"
    df.to_csv(csv_path, index=False)

    # Plot like GeminiOut: habit + phase, adaptive params, dissonance
    plt.figure(figsize=(10,8))

    ax1 = plt.subplot(3,1,1)
    ax1.plot(df.Step, df.K_Strength, label="|K| (rule 1)")
    ax1.plot(df.Step, np.abs(df.Phi_Angle), label="|phase error| (proxy)")
    ax1.set_ylabel("Habit / Phase")
    ax1.legend()

    ax2 = plt.subplot(3,1,2)
    ax2.plot(df.Step, df.Alpha, label="alpha(t)")
    ax2.plot(df.Step, df.Mu, label="mu(t)")
    ax2.set_ylabel("Adaptive params")
    ax2.legend()

    ax3 = plt.subplot(3,1,3)
    ax3.plot(df.Step, df.Dissonance, label="Dissonance")
    ax3.set_ylabel("D")
    ax3.set_xlabel("Step")
    ax3.legend()

    plt.tight_layout()
    png_path = "hd110067_adaptive_trace_py.png"
    plt.savefig(png_path, dpi=160)

    print(f"Wrote {csv_path} and {png_path}")


if __name__ == "__main__":
    main()

File: Code/pit_hd110067_sim/hd110067_adaptive.jl
# hd110067_adaptive.jl
#
# PIT HD 110067 Simulation Engine (Adaptive / Process Fractal)
# Based on Loom Spec v10.1 + Addendum v0.1
#
# Features:
# - Complex K-Field (Phasors)
# - Adaptive Parameters (alpha, mu drift based on Dissonance)
# - Coherence History Tracking
# - "Frozen" Gravity (Baseline) + "Living" Resonance (PIT)
#
# Dependencies:
#   ] add StaticArrays DataFrames CSV
#

using LinearAlgebra
using StaticArrays
using Statistics
using DataFrames
using CSV
using Printf
using Random

# --- 1. ONTOLOGY (Structures) --------------------------------------------

mutable struct Planet
    id::Int
    mass::Float64       # Solar Masses
    pos::MVector{3, Float64} # AU
    vel::MVector{3, Float64} # AU/Day
    name::String
end

mutable struct KernelField
    # The "Habits" (Resonant Phasors for each pair)
    modes::Vector{ComplexF64} 
    
    # Adaptive Parameters (Now Mutable)
    mu::Float64     # Memory Stiffness (Habit Strength)
    nu::Float64     # Novelty (Learning/Plasticity)
    alpha::Float64  # Coupling Strength (PIT Force)
    
    # Adaptive Drift Config (from Addendum)
    k_alpha::Float64 # Drift rate for alpha
    k_mu::Float64    # Drift rate for mu
    d_target::Float64 # Target dissonance (D*)
end

struct ResonanceRule
    inner_id::Int
    outer_id::Int
    p::Int
    q::Int
end

# Coherence History Buffer
mutable struct HistoryBuffer
    window_size::Int
    coherence_log::Vector{Float64} # Rolling window of coherence values
end

const G = 2.95912208286e-4  # AU^3 / (Solar Mass * Day^2)

# --- 2. PHYSICS ENGINE ---------------------------------------------------

function get_lambda(p::Planet)
    x, y = p.pos[1], p.pos[2]
    return atan(y, x)
end

# A. The Interface (Measure Phi)
function measure_phasors(planets::Vector{Planet}, rules::Vector{ResonanceRule})
    phasors = Vector{ComplexF64}(undef, length(rules))
    for (i, rule) in enumerate(rules)
        p_in = planets[rule.inner_id]
        p_out = planets[rule.outer_id]
        
        lambda_in = get_lambda(p_in)
        lambda_out = get_lambda(p_out)
        
        # Resonance Angle: phi = p*λ_out - q*λ_in
        phi = rule.p * lambda_out - rule.q * lambda_in
        phasors[i] = exp(im * phi)
    end
    return phasors
end

# B. Adaptive Parameter Drift (Addendum v0.1)
function adapt_parameters!(K::KernelField, current_dissonance::Float64, dt::Float64)
    # 1. Alpha Drift: Increases when Dissonance is high (Needs stronger correction)
    #    d_alpha = k_alpha * (D - D_target)
    d_alpha = K.k_alpha * (current_dissonance - K.d_target)
    K.alpha += d_alpha * dt
    
    # 2. Mu Drift: Decreases when Dissonance is high (Habit becomes flexible)
    #    d_mu = -k_mu * D
    d_mu = -K.k_mu * current_dissonance
    K.mu += d_mu * dt
    
    # Safety Bounds
    K.alpha = clamp(K.alpha, 1.0e-7, 1.0e-3)
    K.mu    = clamp(K.mu,    0.1,    0.999)
end

# C. The Memory Update (K-Field Evolution)
function update_kernel!(K::KernelField, current_phasors::Vector{ComplexF64}, dt::Float64)
    # Effective Learning Rate depends on Nu and current Saturation
    saturation = abs(mean(K.modes))
    learning_rate = K.nu * (1.0 - saturation)
    
    total_dissonance = 0.0
    
    for i in 1:length(K.modes)
        # Dissonance Vector = Phi - K
        diff = current_phasors[i] - K.modes[i]
        total_dissonance += abs(diff)
        
        # Update K: Move towards Phi
        K.modes[i] += learning_rate * diff * dt
        
        # Renormalize if habit is too strong (Limit |K| <= 1)
        if abs(K.modes[i]) > 1.0
            K.modes[i] = K.modes[i] / abs(K.modes[i])
        end
    end
    
    return total_dissonance / length(K.modes) # Return Avg Dissonance
end

# D. The Manifestation (Forces)
function apply_forces!(planets::Vector{Planet}, star_mass::Float64, K::KernelField, rules::Vector{ResonanceRule})
    # 1. Frozen Habit (Gravity)
    for p in planets
        r_mag = norm(p.pos)
        acc = -G * star_mass * p.pos / (r_mag^3)
        p.vel += acc
    end
    
    # 2. Active Habit (PIT Resonance Torque)
    # Only applies if alpha is significant
    if K.alpha > 1.0e-9
        current_phasors = measure_phasors(planets, rules)
        
        for (i, rule) in enumerate(rules)
            k_mem = K.modes[i]
            z_curr = current_phasors[i]
            
            # Phase Error
            delta_phi = angle(k_mem) - angle(z_curr)
            
            # Torque = Coupling * Memory_Strength * Sin(Error)
            torque = K.alpha * abs(k_mem) * sin(delta_phi)
            
            # Apply tangential kicks
            p_in = planets[rule.inner_id]
            p_out = planets[rule.outer_id]
            
            t_in = cross([0,0,1.0], p_in.pos) / norm(p_in.pos)
            t_out = cross([0,0,1.0], p_out.pos) / norm(p_out.pos)
            
            p_in.vel += (torque / p_in.mass) * t_in
            p_out.vel -= (torque / p_out.mass) * t_out
        end
    end
end

function step_pos!(planets::Vector{Planet}, dt::Float64)
    for p in planets
        p.pos += p.vel * dt
    end
end

# --- 3. MAIN SIMULATION LOOP ---------------------------------------------

function main()
    println("--- PIT ADAPTIVE ENGINE: HD 110067 ---")
    println("Initializing Process Fractal...")
    
    # 1. Init Planets (HD 110067)
    star_mass = 0.81
    m_scale = 3.003e-6
    
    # Init in approximate resonance positions
    planets = [
        Planet(1, 5.69 * m_scale, MVector(0.0793, 0.0, 0.0), MVector(0.0, 0.0, 0.0), "b"),
        Planet(2, 5.69 * m_scale, MVector(0.1039, 0.0, 0.0), MVector(0.0, 0.0, 0.0), "c"),
        Planet(3, 5.69 * m_scale, MVector(0.1364, 0.0, 0.0), MVector(0.0, 0.0, 0.0), "d"),
        Planet(4, 5.69 * m_scale, MVector(0.1790, 0.0, 0.0), MVector(0.0, 0.0, 0.0), "e"),
        Planet(5, 5.69 * m_scale, MVector(0.2166, 0.0, 0.0), MVector(0.0, 0.0, 0.0), "f"),
        Planet(6, 5.69 * m_scale, MVector(0.2621, 0.0, 0.0), MVector(0.0, 0.0, 0.0), "g")
    ]
    
    # Set initial circular velocities + random phase
    Random.seed!(123)
    for p in planets
        v_circ = sqrt(G * star_mass / norm(p.pos))
        theta = rand() * 2 * pi
        r = norm(p.pos)
        p.pos = MVector(r*cos(theta), r*sin(theta), 0.0)
        p.vel = MVector(-v_circ*sin(theta), v_circ*cos(theta), 0.0)
    end

    # 2. Init Kernel (Memory)
    rules = [
        ResonanceRule(1, 2, 3, 2), ResonanceRule(2, 3, 3, 2),
        ResonanceRule(3, 4, 3, 2), ResonanceRule(4, 5, 4, 3),
        ResonanceRule(5, 6, 4, 3)
    ]
    
    # Start with weak, random memory (System hasn't learned yet)
    init_modes = [0.01 * exp(im * rand() * 2 * pi) for _ in 1:length(rules)]
    
    K = KernelField(
        init_modes,
        0.90,   # Initial Mu (High Memory)
        0.05,   # Nu (Learning Rate)
        1.0e-5, # Initial Alpha (Weak Coupling)
        
        # Adaptive Params (ThePage.loom.md)
        1.0e-4, # k_alpha (Drift speed)
        1.0e-5, # k_mu (Drift speed)
        0.10    # D_target (Target dissonance ~ 10%)
    )
    
    println("Kernel Ready. Initial Alpha=$(K.alpha), Mu=$(K.mu)")

    # 3. Run Loop
    dt = 0.1
    steps = 20000
    
    # Logs
    trace_step = Int[]
    trace_phi1 = Float64[]
    trace_K_str = Float64[]
    trace_alpha = Float64[]
    trace_mu = Float64[]
    trace_diss = Float64[]
    
    for t in 1:steps
        # A. Measure Phi
        current_phasors = measure_phasors(planets, rules)
        
        # B. Update K (and get Dissonance)
        avg_diss = update_kernel!(K, current_phasors, dt)
        
        # C. Adaptive Drift (The Living System)
        adapt_parameters!(K, avg_diss, dt)
        
        # D. Apply Forces
        apply_forces!(planets, star_mass, K, rules)
        step_pos!(planets, dt)
        
        # E. Logging
        if t % 20 == 0
            push!(trace_step, t)
            push!(trace_phi1, angle(current_phasors[1]))
            push!(trace_K_str, abs(K.modes[1]))
            push!(trace_alpha, K.alpha)
            push!(trace_mu, K.mu)
            push!(trace_diss, avg_diss)
        end
        
        if t % 5000 == 0
            @printf("Step %d: K-Str=%.3f, Alpha=%.2e, Mu=%.3f, Diss=%.3f\n", 
                    t, abs(K.modes[1]), K.alpha, K.mu, avg_diss)
        end
    end
    
    # 4. Save Output
    df = DataFrame(
        Step = trace_step,
        Phi_Angle = trace_phi1,
        K_Strength = trace_K_str,
        Alpha = trace_alpha,
        Mu = trace_mu,
        Dissonance = trace_diss
    )
    
    filename = "hd110067_adaptive_trace.csv"
    CSV.write(filename, df)
    println("\nSimulation Complete. Data saved to $filename")
end

main()
